### DISCUSSION TITLE
io_uring_wait_cqe_timeout() with a 1ms timeout not returning for seconds

### DISCUSSION THREAD

**Discussion by bateyejoe** [axboe/liburing#810] [Q&A] [answered]
On Ubuntu with 5.19 kernel, I'm testing an app I've converted from using epoll to io_uring. This is actually just a test app for measuring how much capacity is available for my application type. The test app does UDP I/O and simulates sending/receiving many RTP media streams. There is one socket per stream being sent/received, and there is one thread and ring per logical CPU. The media streams are distributed evenly among the rings. Each media stream mimics a g.711 RTP stream having 172-byte UDP payloads sent 20 times a second, spaced as evenly as possible.

Using epoll on this particular hardware, I can manage about 8000 simultaneous streams which is 160,000 packets/s sent and 160,000 packets/s received. My io_uring implementation can barely get about 1000 streams before weird things start happening. In particular, it will run for a short time and then, on all my ring threads, io_uring_wait_cqe_timeout() will block for seconds (usually in the range of 2-5 seconds, but I've seen as high as 20s). As it currently is, I'm passing a hard-coded 1ms as the timeout. Also, when it returns from this very long wait, the return value is zero.

I should add that this is on a physical machine (Intel i7-9700K with an Intel desktop NIC). On another system, which happens to be a hyper-v VM running Ubunti with kernel 6.1, I see better behavior. I can get up to about 4000 streams before io_uring_wait_cqe_time() starts taking too long to reliably send the packets on time (in the range of 300ms when called with a 1ms timeout).

So that's a lot of info and I'm sure I still have issues with my implementation which I'm not looking for help with. I'm Just wondering if anyone has any hints of where to look with regard to the io_uring_wait_cqe_timeout() blocking for so long?

**Comment by redbaron** [comment]
Did you try to get perf report? Might show the reason of a hiccup when taken in that pause.

```
perf record -g
perf report > output_file.txt
```

**Comment by FrankReh** [comment]
Make sure your implementation honors the overflow bit. If the cqeueue had an overflow, you have to "manually" get it to read from the overflow area first before normal operation will proceed.

**Reply by bateyejoe** [reply]
No, I'm not doing anything of this sort. But I've googled around for more information on the subject but I'm having trouble finding much information. If you have any links to places that explain this further I would greatly appreciate it.

**Comment by axboe** [comment]
Since it sounds it's stuck, might also be worth doing:

```
# cat /proc/$pid/wchan
```

where $pid is the pid of the task as well. And you can capture the ring state by catting fdinfo/$fd  in that same /proc/$pid/ directory, with $fd being the fd value of the ring fd, which might also provide some hints.

This isn't normal, and I'm curious what's going on here.

**Reply by bateyejoe** [reply]
Watching the fdinfo, I see the SQEs are almost always 0, but the CQEs go in cyles from 0 up to 1024. Seems like things are getting backed up and then completing all at once, but sort of goes in cycles. Watching the bandwidth on the network monitor shows something similar where it runs perfect for a few seconds, drops to almost zero, shoots up higher than the normal steady-state and begins to roughly cycle like that. I've created a stand-alone app that reproduces the issue. It's about a 1000 lines of code. Should I just post it here in the chat?

**Reply by axboe** [reply]
> Should I just post it here in the chat?

Please do!

**Reply by bateyejoe** [reply]
The only dependencies are the c++ standard library and liburing so should be straightforward to compile. For reproducing this, I'm running two copies of the app on two different machine. One side is the send side which produces the packets at the configured rates, the other side is a reflector that simply sends received packets back to the originator. For simplicity, both instances need to be using the same port numbers, so you can't really run both on one machine. 

Typically, I'm running the 'send' mode of this app against my epoll reflector app, so running this against the uring version on the other system may produce worse results since they may both experience problems. For my i7-9700k machine, I'm running with the following command:

msim send 192.168.0.208 50000 1000 50 8 10

That's the dest IP, starting port of 50,000, 1000 sockets (on ports 50,000 through 50,999) 50 packets per second, 8 rings/thread (same as number of cores), and 10 seconds run time. I can get this to run stable (about 10MB/s in both directions), but when I bump it up to 2000 sockets, it starts having the problems blocking waiting for a cqe. If you don't want to use as many sockets, you can get the same packet rate by bumping up the packets pers second to 500 and dropping the sockets to 100 or 200, e.g.

msim send 192.168.0.208 50000 100 500 8 10

10 seconds is usually plenty of time to reproduce the issue. On the other system, I would run the reflector mode with plenty of sockets and plenty of duration:

msim reflect 50000 3000 8 600

That creates 3000 sockets starting at port 50,000, 8 rings and 10 minute duration. Ideally, use a beefier machine for the reflector so you can troubleshoot one instance at a time.

**Reply by bateyejoe** [reply]
[msim.zip](https://github.com/axboe/liburing/files/10946506/msim.zip)
It seemed to lost a lot of code, even using the 'code' escapes. So I just zipped it up.

**Reply by bateyejoe** [reply]
I wrote a new test to try and narrow down this issue. Instead of lots of low-rate streams on lots of socket on a ring/cpu, I'm doing a single high rate stream (still with small 172 byte UDP payloads) on a single socket on a single ring/thread. This test can't reproduce the multi-second blocking I saw in io_uring_wait_cqe_timeout(), but it does highlight the large disparity in CPU consumption of io_uring vs epoll. I also observe that the more receive operations posted to the ring at any given time, the more the CPU consumption for the same packet rate. In other words, it seems the number of operations in-flight in the ring has a significant effect on CPU usage, even for the same I/O completion rates. I've posted my test apps and some summarized results.
[AsyncUdpTests](https://github.com/bateyejoe/AsyncUdpTests)

**Reply by axboe** [reply]
@bateyejoe thanks for the writeup and reproducer, I'll try and take a look next week. The only thing I can think of that depends on the number of pending requests is the poll callback, as each pending request will have one armed. But that really should not be different from epoll, this is how both of them get notified after all.

Maybe I missed it when I skimmed your link, but what kernel version are you running?

**Reply by bateyejoe** [reply]
@axboe No problem. Seems like that could be a big difference, especially in this test case. Since this test case has only one socket, there is only one file descriptor registered with epoll and it only has two events (in and out) enabled. But even in my real-world app, I would typically have, for each socket, 8 pending receives and maybe one to two pending sends whereas the epoll version would be registered for two events per socket.

I also think there must be some significance to the increasing amount of time io_uring_submit() takes as the number of outstanding ring ops increases. The test app has a simple profiler framework that times all the io_uring_xxx() calls and summarizes them at the end. There's also an option to record the duration and submission batch size of every call to io_uring_submit() in memory, and then write to out at the end. I didn't really see a correlation between the batch size and the duration of the io_uring_submit() call. It just seemed to be generally longer, the more ring ops were in flight.

**Reply by axboe** [reply]
One immediate thought, particularly for one socket, is that you should be using multi-shot rather than single shot. I haven't looked at your code yet, but I'm assuming you're not. For a single socket, if you keep epoll armed but do separate receives with io_uring, then that's really a worst case kind of setup. With multishot on the socket and using provided ring buffers, then that'd be a much better architecture for a setup where you have N sockets that continually send and receive (or even just receive, obviously).

But I'll poke more next week and try and run what you have.

**Reply by axboe** [reply]
Also see: https://github.com/axboe/liburing/wiki/io_uring-and-networking-in-2023

**Reply by bateyejoe** [reply]
No, I'm not using multi-shot because it would be difficult to use in my real-world app. In short, I have a lot of buffers, not all of which are used in I/O and I'm not sure about having so many buffers registered with io_uring. It is something I plan to investigate as an improvement if I can get io_uring to be at least as efficient as epoll with the current design.

Also, I think there might be a misconception about the included epoll implementation. The epoll implementation does not use io_uring at all, it's pure epoll, and the io_uring version does not use epoll at all.

Thanks for looking at this. I may play around with adding multishot support next week.

**Reply by axboe** [reply]
I don't think you'll get io_uring to be as efficient if you're doing thousands of requests to the same socket rather than just using multishot. The latter would most likely get there, but your existing approach is not likely to scale very well. We could obviously abstract this out a bit and have a shared queue sitting on top of the socket, but I don't think it's going to help a lot of real world cases. I did play with that at one point to basically keep polled armed even for single shot, but I'm not entirely sure it's workable. Multi-shot is really the way to do this with io_uring.

I didn't make any assumptions about the epoll backend using io_uring, unsure where that is from?

**Reply by bateyejoe** [reply]
I think I misunderstood when you said "if you keep epoll armed but do separate receives with io_uring", I thought you were talking about using epoll together with io_uring. Now maybe you were comparing the two schemes. 

I do have a concern about multishot. If, for example, I have this media application where I expect all received packets to be around the same small size, I could create a big pool (presumably, one for each ring) and share this pool among all my sockets on that ring. But this means that one rogue sender, sending to one of the sockets, could exhaust most or all of the available buffers by simply flooding one socket with packets. That's ok if it affects only that socket, but it could starve every other socket using the buffer group, bringing down a lot of streams.  In my application, I deal with this by simply having a small number of receive buffers per socket (usually 8). If those are all used, and the socket layer's buffering fills up, packets for that socket are simply dropped, which is fine. Having to monitor the network pattern for bad actors is a much more difficult problem and it's not even clear if I could react fast enough to prevent it from affecting other sockets.

Would it be realistic to have a different buffer group id for every socket and just have a small number of buffers in each buffer group? I'm talking about having thousands of sockets, so it would require thousands of buffer groups.

**Reply by bateyejoe** [reply]
So I added multishot to my test, but as part of that, I upgraded my kernal to 6.00.19. The multishot test shows much better performance, but so does the non-multishot test. Apparently, there were some significant internal improvements (at least for my use case) between 5.19 and 6.0. I'm trying to get some specific numbers, but I can't get the test to complete because of a different issue.

The reason my tests aren't completing is that a io_uring_prep_cancel_fd(sqe, sockFd, IORING_ASYNC_CANCEL_FD | IORING_ASYNC_CANCEL_ALL) is only cancelling a few of the outstanding receive operations (I've seen between 2 and 5 ops cancelled), no matter how many are currently in flight for that file descriptor. This worked perfectly in 5.19. I even added logic to re-submit the cancel, but that doesn't result in any other cancelled receives. I get the CQE for the cancel operation itself (usually first) and the result is 0. Here is a dump of fdinfo for the ring after cancel only cancelled 5 of the 2000 outstanding receive operations pending on the socket's fd:

```
kevino@sim2:~$ cat /proc/2902776/fdinfo/4
pos:	0
flags:	02000002
mnt_id:	15
ino:	389343
SqMask:	0xfff
SqHead:	4145363
SqTail:	4145363
CachedSqHead:	4145363
CqMask:	0x1fff
CqHead:	4143368
CqTail:	4143368
CachedCqTail:	4143368
SQEs:	0
CQEs:	0
SqThread:	-1
SqThreadCpu:	-1
UserFiles:	1
    0: UDP
UserBufs:	0
PollList:
CqOverflowList:

```

I'm pretty sure that my other receives have never completed. Is this possibly a known bug in this kernel version?

**Reply by axboe** [reply]
Not aware of such a bug, bug wrote a quick test case and it seems to work for me. Used current tree and 6.1-stable. Can you give it a spin with your kernel? It should just print each cqe we get. If it hangs before completing, then it's missing some completions.

```
/* SPDX-License-Identifier: MIT */
/*
 * Description: check that CANCEL_ALL on the fd finds all
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/socket.h>

#include <liburing.h>

#define NREQS	128

static int test(struct io_uring *ring)
{
	struct io_uring_sqe *sqe;
	struct io_uring_cqe *cqe;
	int fd[2], ret, i;
	char buf[64];

	if (socketpair(PF_LOCAL, SOCK_STREAM, 0, fd) < 0) {
		perror("socketpair");
		return 1;
	}

	for (i = 0; i < NREQS; i++) {
		sqe = io_uring_get_sqe(ring);
		io_uring_prep_recv(sqe, fd[0], buf, sizeof(buf), 0);
		sqe->user_data = 1;
	}

	ret = io_uring_submit(ring);
	if (ret != NREQS) {
		fprintf(stderr, "submit %d\n", ret);
		return 1;
	}

	sqe = io_uring_get_sqe(ring);
	io_uring_prep_cancel_fd(sqe, fd[0], IORING_ASYNC_CANCEL_ALL);
	sqe->user_data = 2;

	ret = io_uring_submit(ring);
	if (ret != 1) {
		fprintf(stderr, "submit %d\n", ret);
		return 1;
	}

	for (i = 0; i < NREQS + 1; i++) {
		ret = io_uring_wait_cqe(ring, &cqe);
		if (ret) {
			fprintf(stderr, "cqe wait %d\n", ret);
			return 1;
		}
		printf("req%d: data=%ld, res=%d\n", i, (long) cqe->user_data, cqe->res);
		io_uring_cqe_seen(ring, cqe);
	}

	close(fd[0]);
	close(fd[1]);
	return 0;
}

int main(int argc, char *argv[])
{
	struct io_uring ring;
	int ret;

	io_uring_queue_init(NREQS, &ring, 0);

	ret = test(&ring);
	if (ret) {
		fprintf(stderr, "Test failed\n");
		return ret;
	}

	return 0;
}
```

**Reply by bateyejoe** [reply]
That works for me. I've been trying to narrow it down. Originally seeing it in my test app at 200000 packets/s test with 2000 outstanding receives. So I reduced everything and I can get it to sporadically happen at 5000 packets/s and 1600 outstanding receives. So perhaps this is a logic error in my app, let me investigate further.

**Reply by bateyejoe** [reply]
I modified your sample for UDP and added some incoming packets and it pretty reliably reproduces the problem. If you have any problems reproducing this, comment out the nanosleep() in sender_thread() to increase the send rate. The higher the rate, the more receives that don't get cancelled. I was able to reproduce this with as few as 128 receive ops, but it happened more reliably the more receive ops there are, so I made that a command line param.


```
/* SPDX-License-Identifier: MIT */
/*
 * Description: check that CANCEL_ALL on the fd finds all
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <pthread.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include <liburing.h>

unsigned int g_receiveTarget = 10000;
unsigned int g_numReceives = 128;

static int test(struct io_uring *ring, int fd)
{
	struct io_uring_sqe *sqe;
	struct io_uring_cqe *cqe;
	int ret, i;
	char buf[64];

    unsigned int inFlight = 0;
    unsigned int recvCompletions = 0;

    printf("Initial submit of %u receive operations.\n", g_numReceives);

	for (i = 0; i < g_numReceives; i++) {
		sqe = io_uring_get_sqe(ring);
		io_uring_prep_recv(sqe, fd, buf, sizeof(buf), 0);
		sqe->user_data = 1;
        ++inFlight;
	}

	ret = io_uring_submit(ring);
	if (ret != g_numReceives) {
		fprintf(stderr, "submit %d\n", ret);
		return 1;
	}

    bool cancelled = false;

    while (inFlight != 0)
    {
		ret = io_uring_wait_cqe(ring, &cqe);
		if (ret) {
			fprintf(stderr, "cqe wait %d\n", ret);
			return 1;
		}
//		printf("req%d: data=%ld, res=%d\n", i, (long) cqe->user_data, cqe->res);
        --inFlight;
        if (cqe->user_data == 1)
        {
            ++recvCompletions;
            // Handle receive completion
            if (cqe->res < 0)
            {
                printf("Receive result %d, inFlight %d\n", cqe->res, inFlight);
            }
            else if ((recvCompletions % 1000) == 0)
            {
                printf("Receive completions: %d\n", recvCompletions);
            }

            if (recvCompletions < g_receiveTarget)
            {
                sqe = io_uring_get_sqe(ring);
                io_uring_prep_recv(sqe, fd, buf, sizeof(buf), 0);
                sqe->user_data = 1;
                ret = io_uring_submit(ring);
                if (ret != 1) {
                    fprintf(stderr, "submit %d\n", ret);
                    return 1;
                }
                ++inFlight;
            }
            else if (!cancelled)
            {
                printf("Max receives reached, queueing cancel\n");
                sqe = io_uring_get_sqe(ring);
                io_uring_prep_cancel_fd(sqe, fd, IORING_ASYNC_CANCEL_ALL);
                sqe->user_data = 2;
                cancelled = true;
                ret = io_uring_submit(ring);
                if (ret != 1) {
                    fprintf(stderr, "submit %d\n", ret);
                    return 1;
                }
            }
        }
        else if (cqe->user_data == 2)
        {
            // Handle Cancel
            printf("Cancel complete res=%d\n", cqe->res);
        }
		io_uring_cqe_seen(ring, cqe);
	}

	return 0;
}

static void* sender_thread(void* arg)
{
    int fd = (int)(uintptr_t)arg;
    char buf[64];
    struct timespec sleepTime;

    sleepTime.tv_sec = 0;
    sleepTime.tv_nsec = 1000;   // Send about once every 0.1 ms

    for (unsigned int i = 0; i < g_receiveTarget+10; i++)
    {
        if (send(fd, buf, sizeof(buf), 0) != sizeof(buf))
        {
            fprintf(stderr, "Send failed. %d\n", errno);
            break;
        }
        /// Take out this sleep to make the problem even more reproducible
        nanosleep(&sleepTime, NULL);        
    }

    return 0;
}

int main(int argc, char *argv[])
{
	struct io_uring ring;
	int ret;
    int fd[2];
    pthread_t threadId;
    socklen_t sockLen;
    struct sockaddr_in addr0;
    struct sockaddr_in addr1;

    if (argc > 1)
    {
        g_numReceives = atoi(argv[1]);
    }

	io_uring_queue_init(g_numReceives, &ring, 0);

    // Local addresses
    addr0.sin_family = AF_INET;
    addr0.sin_port = 0;
    inet_aton("127.0.0.1", (struct in_addr*)&addr0.sin_addr.s_addr);
    addr1.sin_family = AF_INET;
    addr1.sin_port = 0;
    inet_aton("127.0.0.1", (struct in_addr*)&addr1.sin_addr.s_addr);

    fd[0] = socket(AF_INET, SOCK_DGRAM, 0);
    if (fd[0] == -1)
    {
        perror("socket");
        return 1;
    }
    if (bind(fd[0], (const struct sockaddr*)&addr0, sizeof(addr0)))
    {
        perror("bind");
        return 1;
    }
    sockLen = sizeof(addr0);
    if (getsockname(fd[0], (struct sockaddr*)&addr0, &sockLen))
    {
        perror("getsockname");
    }

    fd[1] = socket(AF_INET, SOCK_DGRAM, 0);
    if (fd[1] == -1)
    {
        perror("socket");
        return 1;
    }
    if (bind(fd[1], (const struct sockaddr*)&addr1, sizeof(addr1)))
    {
        perror("bind");
        return 1;
    }
    if (getsockname(fd[1], (struct sockaddr*)&addr1, &sockLen))
    {
        perror("getsockname");
        return 1;
    }

    if (connect(fd[0], (const struct sockaddr*)&addr1, sizeof(addr1)))
    {
        perror("connect");
        return 1;
    }
    if (connect(fd[1], (const struct sockaddr*)&addr0, sizeof(addr0)))
    {
        perror("connect");
        return 1;
    }

//  if (socketpair(PF_LOCAL, SOCK_STREAM, 0, fd) < 0) {
//		perror("socketpair");
//		return 1;
//	}

    ret = pthread_create(&threadId, NULL, &sender_thread, (void*)(uintptr_t)fd[1]);
    if (ret != 0)
    {
        fprintf(stderr, "Failed to create sender thread. %d\n", ret);
    }

	ret = test(&ring, fd[0]);
	if (ret) {
		fprintf(stderr, "Test failed\n");
		return ret;
	}

    close(fd[0]);
    close(fd[1]);

	return 0;
}

```

**Reply by axboe** [reply]
Excellent, I'll give that a go!

Just to clarify, you said this was never an issue with 5.19, but with 6.0 it is? Gives me a range of things to look through.

**Reply by axboe** [reply]
I'm not seeing any hangs off cancelation, but I am seeing it stall ala:

```
axboe@m1max-kvm ~> ./cancel-udp
Initial submit of 128 receive operations.
Receive completions: 1000
Receive completions: 2000
Receive completions: 3000
Receive completions: 4000
Receive completions: 5000
Receive completions: 6000
Receive completions: 7000
Receive completions: 8000
Receive completions: 9000
```

and looking at stats, we have had 9619 submissions and 9419 completions. That's a difference of 128. This matches the SQ head and CQ tail as well. Looking at fdinfo, there are 128 recv pending, waiting in poll.

While this does look odd, unless things got dropped on the sending side, it looks different to what you are talking about which seems to be cancel that doesn't find all the pending requests...

**Reply by bateyejoe** [reply]
You're apparently not reaching the cancel operation. Since it's UDP, it may drop some sends. I added 10 to the sends in the for loop for the sender_thread(), you could make that much higher to make sure you get enough receives to begin the shutdown logic.

Looks like you're pretty close since it triggers at 10000.

**Reply by bateyejoe** [reply]
And that's correct about the versions, it worked perfectly for me in 5.19 and now I'm running on 6.0.19 (just picked that because I believe multishot recvmsg requires 6.0 and 6.0.19 was the highest 6.0 version).

**Reply by axboe** [reply]
Can't reproduce a cancel hang once I bump send count. I'll go back to 6.1-stable and see if it hits anything.

**Reply by bateyejoe** [reply]
Might try taking out the nanosleep() in the sender_thread() (if you haven't already) and bumping up the number of receive operations by passing it on the command line seems to make it happen more reliably. I would try 1000 or 2000 receive operations.

**Reply by bateyejoe** [reply]
With the nanosleep(), I'm only missing cancels on about 1-4 buffers (or none at all). Without the nanosleep, and with 1000 receive ops, I'm missing cancels on around 150 buffers.

**Reply by axboe** [reply]
Took a closer look and this is specific to 6.0, newer should work fine. This is unfortunately not something I can fix since 6.0-stable is long since retired, it's not one of the longer term maintained kernels. 6.1-stable and 6.4-stable are current, I would strongly recommend upgrading to that.

**Reply by bateyejoe** [reply]
Moved up to 6.1.39 and the cancel is finding all the operations and cancelling them correctly. Multishot is working and is much better than the individual receive operations, but still not quite as good as epoll, but pretty close. In the pic below, the first run (on the left) is io_uring using multishot recvmsg, and the second (on the right) is epoll. These are both running at 465000 packets/s which is essentially saturating my 1Gbps link between the two systems.
![image](https://github.com/axboe/liburing/assets/32120141/68d43ff9-efbc-4acb-8ba0-786313954ac8)

On 6.1 the non-multishot version still performs pretty poorly. I recall that it looked like it was performing much better on 6.0.19, but I'm not going to bother going back to re-test since that kernel has the cancel problem. Multishot is the only way to go with this kind of traffic.

On the subject of the ring buffer, I do have a question. If I register a ring buffer, fill it with buffers, and use it until I cancel my multi-shot receive, what's the typical way to reclaim the un-used buffers from the buffer ring? I see that I can un-register the buffer ring from the io_uring, but I don't see anything for accessing the unused buffers left in the ring.

**Reply by bateyejoe** [reply, marked as answer]
So, to summarize, the end result of all this is that the problems I was seeing on 5.19 where I could block in io_uring_wait_cqe_timeout() for multiple seconds with a 1ms timeout, are all gone in 6.1.39. I'm running the same build of my original test app that had the problems in 5.19 under 6.1.39 and I'm not seeing any extended blocking in the wait and performance is close to epoll, even when not using multishot receives. So apparently there were some internal improvements between 5.19 and 6.1.

**Comment by isilence** [comment]
And having a reproducer would also be great

**Comment by FrankReh** [comment]
Where is the overflow bit being handled in that msim.cpp file? Grep didn't find one for me. If you aren't handling the overflow case, that is still the first place I would look for the cause. Sorry if my armchair quarterbacking is just noise here.

**Reply by bateyejoe** [reply]
Not noise at all. I just can't find much info on handling the overflow bit. Incidentally, I just tried adding code for checking for the overflow bit before each call to io_uring_wait_cqe_timeout() and if set, calling io_uring_get_events(), but in my test run that reproduces this issue, the io_uring_cq_has_overflow() never returns non-zero.

Also, my test runs ok, on this hardware, at 1000 streams/sockets with the ring configured for 512 entries. My test fails at 2000 streams/sockets and it doesn't seem to matter if I bump the ring's queue entries up to 2048.

**Comment by FrankReh** [comment]
Checking, and handling the case when there is a cqueue overflow bit was the critical piece for some software we worked on months ago that exhibited choppy throughput. The overflow bit is described in the manuals, I get 30 hits in liburing/man/* but of course some are soft link duplicated files.

Probably the man pages describe this better than I can.
But if/when you call the `enter` syscall, you should be including the IORING_ENTER_GETEVENTS in `flags` when there is an overflow. I guess it is possible user code *always* sets that bit, I didn't check that in your example, in which case it should not be necessary to check for the overflow because all you can do about it is call `enter` with the GETEVENTS bit set anyway.
